### PR TITLE
Updates to diagram properties dialog

### LINKED ISSUE #4458: [FEATURE] Add spacing option for vector layer bar chart diagrams
Original commit: https://github.com/qgis/QGIS/commit/49dcc2c253536f23d199afa215570bde3aa1b582 by nyalldawson

Allows for user-controlled spacing between each bar in the chart.

Sponsored by SLYR

### LINKED ISSUE #4464: [FEATURE][diagrams] Add option to control pie diagram angular direction
Original commit: https://github.com/qgis/QGIS/commit/baf38199e09651f7d85914b80fc1cc3cc64d2d3a by nyalldawson

I.e. control whether the sections are rendered in a clockwise or
anticlockwise direction

Sponsored by SLYR

### LINKED ISSUE #4469: [FEATURE][diagrams] Add option to show diagram axis for histogram diagram symbols
Original commit: https://github.com/qgis/QGIS/commit/9a23e1b20946a5315c1522ccdf984cac9663b5f8 by nyalldawson

Where the axis line symbol can be set using a standard QGIS line symbol.

Sponsored by SLYR

### LINKED ISSUE #4475: [FEATURE][diagrams] New diagram type "stacked bars"
Original commit: https://github.com/qgis/QGIS/commit/204bd47b19b096ebb13b8d943f875fd5e42443be by nyalldawson

Stacks bars of varying colors for each attribute on top of each other
vertically or horizontally.

Sponsored by SLYR

### LINKED ISSUE #4478: [FEATURE][diagrams] Paint effect support for diagram renderer
Original commit: https://github.com/qgis/QGIS/commit/e8ec0042f4cb3e9c67d085b583d895daafbcdad9 by nyalldawson

Allows for diagrams to use paint effects, including drop shadows,
outer glows, etc...

Sponsored by SLYR

### PR DESCRIPTION
**Pull request by DelazJ**
Fixes #4458, Fixes #4464, Fixes #4469, Fixes #4475, Fixes #4478

### CODE REVIEW

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 1]
```diff
@@ -1663,12 +1666,18 @@ Rendering
...
 * the graphic's opacity, its outline width and color;
-* and, depending on the type of diagram:
-
-  * the width of the bar in case of histogram;
-  * the circle background color in case of text diagram, and the font used for texts;
-  * the orientation of the left line of the first slice represented in pie chart.
-    Note that slices are displayed clockwise.
+* depending on the type of diagram:
+
+  * in case of histogram and stacked bars, the width of the bar and the spacing
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
  * for histogram and stacked bars, the width of the bar and the spacing
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 2]
```diff
@@ -1663,12 +1666,18 @@ Rendering
...
+    between the bars. You may want to set the spacing to ``0`` for stacked bars.
+    Moreover, the :guilabel:`Axis line symbol` can be made visible in the
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
    Moreover, the :guilabel:`Axis line symbol` can be made visible on the
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 3]
```diff
@@ -1663,12 +1666,18 @@ Rendering
...
+    map canvas and customized using :ref:`line symbol properties
+    <vector_line_symbols>`.
+  * in case of text diagram, the circle background color and
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
  * for text diagram, the circle background color and 
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 4]
```diff
@@ -1663,12 +1666,18 @@ Rendering
...
+* depending on the type of diagram:
+
+  * in case of histogram and stacked bars, the width of the bar and the spacing
+    between the bars. You may want to set the spacing to ``0`` for stacked bars.
+    Moreover, the :guilabel:`Axis line symbol` can be made visible in the
+    map canvas and customized using :ref:`line symbol properties
+    <vector_line_symbols>`.
+  * in case of text diagram, the circle background color and 
+    the :ref:`font <font_selector>` used for texts;
+  * in case of pie charts, the :guilabel:`Start angle` of the first
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
  * for pie charts, the :guilabel:`Start angle` of the first
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 5]
```diff
@@ -1700,15 +1708,26 @@ Size
...
+'Points', 'Pixels', 'Map Units' or 'Inches'.
 You can use :
 
-* :guilabel:`Fixed size`, an unique size to represent the graphic of all the
-  features, except when displaying histogram
-* or :guilabel:`Scaled size`, based on an expression using layer attributes.
-
-.. ToDo: better explain the scale behaviour of the different diagram types
+* :guilabel:`Fixed size`, a unique size to represent the graphic of all the
+  features (not available for histogram)
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
  features (not available for histograms)
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 6]
```diff
@@ -1700,15 +1708,26 @@ Size
...
+* or :guilabel:`Scaled size`, based on an expression using layer attributes:
+
+  #. In :guilabel:`Attribute`, select a field or build an expression
+  #. Press :guilabel:`Find` to return the :guilabel:`Maximum value` of the
+     set attribute or enter a custom value in the widget.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
     attribute or enter a custom value in the widget.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 7]
```diff
@@ -1700,15 +1708,26 @@ Size
...
+  #. For histogram and stacked bars, enter a :guilabel:`Bar length` value,
+     used to represent the :guilabel:`Maximum value` of the attributes.
+     For each feature, the bar lenght will then be scaled linearly to keep
+     this matching.
+  #. For pie chart and text diagram, enter a :guilabel:`Size` value,
+     used to represent the :guilabel:`Maximum value` of the attributes.
+     For each feature, the circle area or diameter will then be scaled
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 8]
```diff
@@ -1700,15 +1708,26 @@ Size
...
+     linearly to keep this matching (from ``0``).
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 9]
```diff
@@ -1723,25 +1742,28 @@ Placement
 ---------
 
 :guilabel:`Placement` helps to define diagram position.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
:guilabel:`Placement` defines the diagram position.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 10]
```diff
@@ -1723,25 +1742,28 @@ Placement
...
-* 'Over the centroid', 'Around the centroid' (with a distance set),
-  'Perimeter' and anywhere 'Inside polygon' are the options for polygon features.
+* :guilabel:`Around centroid' (with a distance set)`, :guilabel:`Over centroid`,
+  :guilabel:`Using perimeter` and :guilabel:`Inside polygon`
+  are the options for polygon features.
 
-The diagram can also be placed using feature data to fill the coordinates ``X``
-and ``Y`` fields.
+The diagram can also be placed using feature data to fill the :guilabel:`X`
+and :guilabel:`Y` coordinate fields.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]: What does "using feature data" mean?
Should there be an example?

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 11]
```diff
@@ -1755,10 +1777,11 @@ the **Priority** slider value.
 Options
 -------
 
-The :guilabel:`Options` tab has settings only in case of histogram. You can
-choose whether the bar orientation should be 'Up', 'Down', 'Right' and 'Left'.
+The :guilabel:`Options` tab has settings for histogram and stacked bars.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
The :guilabel:`Options` tab has settings for histograms and stacked bars.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 12]
```diff
@@ -1755,10 +1777,11 @@ the **Priority** slider value.
...
+You can choose whether the :guilabel:`Bar orientation` should be
+:guilabel:`Up`, :guilabel:`Down`, :guilabel:`Right` or :guilabel:`Left`,
+allowing to generate horizontal or vertical diagrams.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
for horizontal and vertical diagrams.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst (hunk 7 above)
Comment (DelazJ) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: I disagree with this removal. It addresses a setting in tghe GUI, explains how the chart are calsulated and follows the same pattern as the previous point

File: source/docs/user_manual/working_with_vector/vector_properties.rst (hunk 10 above)
Comment (DelazJ) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: euh... data-defined feature?

Comment (havatv) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: Yes. I have not checked the GUI, but something like:
```
The diagram can also be placed by data-defining the :guilabel:`X` and :guilabel:`Y`
coordinate fields.
```

**Pull request merged**